test.py: add cluster and new approval tests

kostja opened the pull request:
The purpose of this series is to introduce infrastructure
for managed scylla processes into test.py,
switch some existing suites to use test.py managed processes
and introduce cluster tests.

All of this is expected to make possible to test Raft topology
changes and schema changes using an easy to use and fast tool
such as test.py. In general this will allow testing Scylla clusters
from within the development test harness.

Branch URL: kostja/test.py.v5

Changes in v6:
* properly cancel unused artifacts
* properly delete old tmpdirs
* don't accidentally cleanup wrong suite artifacts in multi-mode testing
* invoke suite exit artifacts ASAP, without waiting for exit 
* add pre- and post-check for cql-pytest
* accept str type in extra_cmdline_options
* write test name markers into server log
* don't drop scylla server logs if there is at least one failed test
* grab the server logs on failure and print them out
* workaround a bug in python event loop polluting the console output on exception
* add test pre-check and post-checks for tests running against shared scylla instances to make sure 
such tests don't leave garbage around
* introduce ScyllaCluster object to represent a cluster of scylla servers
* fix "coroutine was never awaited" errors on exceptions
* add unidiff output to the log

Changes in v5:
* misc nits x2
* fix build failure (update lwt_list.test)
* capture server logs on failure
* suppress noise logging from cassandra-python

Changes in v3:

* type annotations
* more aggressive use of pathlib
* drop alternator suite
* switch approval tests to Python

Changes in v2:

* cluster tests
* short license headers
* dropped move of run.py
* dropped use of taskset, not useful in cluster/pooled environment, --overprovisioned is on anyway
* cleaned up scylla server options
* removed the switch for alternator tests
* style fixes
* use async http request and cql connect to check that the server is up
* use more pathlib

kostja commented:
Testing jenkins. Somehow the build pipeline still doesn't have the required dependencies.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/140/

nyh commented on test.py:
```diff
@@ -936,11 +936,11 @@ def read_log(log_filename):
         print("The following test(s) have failed: {}".format(
             palette.path(" ".join([t.name for t in failed_tests]))))
-        if not output_is_a_tty:
+        if options.verbose:
```
I think this change is orthogonal to this series. @xemul did this output_is_a_tty thing because he explicitly didn't want to see long output on the terminal, and only see it when the output is redirected (as it is on Jenkins, or even if you just pipe it to "less"). I agree that it's weird that we have an explicit "--verbose" option and ignore it in this case, but I am not sure what were the original considerations. For example, it's easier to type "test.py | less" to see the full output instead of "test.py --verbose | less" which you'll need after your patch.
I would like @xemul to review this change.

xemul replied:
I understand the justification from @kostja , but I still don't see the use of the output of any test anywhere but in this test's log file. Even less I see the point in printing test log on the screen. The "test.py --verbose | less" seems very artificial, it will work even without this new --verbose option (let alone the fact that you have test log file anyway, why not lessing it?).

xemul replied:
By and large this change makes sense to me, I don't mind having it merged.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/167/

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+        self.cfg["seeds"] = self.cfg.get("seeds") or self.cfg["host"]
+        # Use the last part in host IP 127.151.3.27 -> 27
+        # There can be no duplicates within the same test run
+        # thanks to how host registry registers subnets, and
```
Can you specify it in `HostRegistry` (in the docstring at class definition)? Now it's a trick that depends on internal implementation detail of `HostRegistry`, instead it should be specified at the interface level.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+        # There can be no duplicates within the same test run
+        # thanks to how host registry registers subnets, and
+        # different runs use different vardirs.
+        self.shortname = pathlib.Path("scylla-" + self.cfg["host"].split(".")[-1])
```
it seems that `self.shortname` could be a local variable (not member)

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+
+        self.log_file_name = self.vardir / self.shortname.with_suffix(".log")
+
+        self.config_file_name = self.cfg["workdir"] / "conf/scylla.yaml"
```
`/ 'conf' / 'scylla.yaml'` would be more elegant

kostja replied:
No it wouldn't, I like it more this way.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+
+        # Cleanup any remains of the previously running server in this path
+        shutil.rmtree(self.tmpdir, ignore_errors=True)
+        pathlib.Path(self.cfg["workdir"]).unlink(missing_ok=True)
```
You don't need to construct `Path` again, `self.cfg['workdir']` is already a `Path`.
Same for the 3 lines below.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+        pathlib.Path(self.cfg["workdir"]).symlink_to(self.tmpdir)
+        pathlib.Path(self.config_file_name).parent.mkdir(parents=True, exist_ok=True)
+        # Create a configuration file.
+        with open(self.config_file_name, 'w') as config_file:
```
`self.config_file_name.open('w')`

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+        with open(self.config_file_name, 'w') as config_file:
+            config_file.write(SCYLLA_CONF_TEMPLATE.format(**self.cfg))
+
+        self.log_file = open(self.log_file_name, "wb")
```
`self.log_file_name.open('wb')`

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+
+        self.log_file = open(self.log_file_name, "wb")
+
+    def find_log_file_pattern(self, fil, pattern_re) -> bool:
```
Could be a free-function or static method since it doesn't use `self`
But is it used anywhere?

kostja replied:
Not any more. But it might be used in the future.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+
+class ScyllaServer:
+    # Regular expression to scan the log
+    STARTUP_MSG_RE = re.compile(".*Scylla.*initialization completed")
```
Used anywhere?

kostja replied:
Not any more.

kostja replied:
Removed.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+        auth = PlainTextAuthProvider(username='cassandra', password='cassandra')
+        try:
+            with Cluster(contact_points=[self.cfg["host"]], auth_provider=auth) as cluster:
+                with cluster.connect():
```
Are we fine with this blocking? (not sure how the async code interacts with driver code)

kostja replied:
It is blocking. For now we're fine with it, yes.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+        # Add suite-specific command line options
+        scylla_args = SCYLLA_CMDLINE_OPTIONS + self.cmdline_options
+        env = os.environ.copy()
+        env.clear()     # pass empty env to make user user's SCYLLA_HOME has no impact
```
If we `clear()` it, why do we `copy()` the `os.environ`? It seems like we simply need an empty `env` from the beginning.

kostja replied:
I explained in a reply to Nadav. There is no other type-safe way to get one.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+            preexec_fn=os.setsid,
+        )
+
+        self.start_time = time.time()
```
Seems like it could be a local variable

kostja replied:
It might be useful as a non-local one in the future. Overall I don't think it matters.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+
+        while time.time() < self.start_time + START_TIMEOUT:
+            if self.cmd.returncode:
+                with open(self.log_file_name, 'r') as log_file:
```
`self.log_file_name.open('r')`

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+            with cluster.connect() as session:
+                session.execute("CREATE KEYSPACE k WITH REPLICATION = {" +
+                                "'class' : 'SimpleStrategy', 'replication_factor' : 1 }")
+                session.execute("DROP KEYSPACE k")
```
Interesting. Does this wait for the schema change to propagate to the entire cluster?

kbr- replied:
By the way, even if it did wait, I don't think this fixes the race between token propagation and streaming.

To fix that you would have to force gossip propagation in the middle of bootstrap, **after** new tokens were inserted by joining nodes but **before** streaming starts. There is no reliable way to do that. Right now you call `force...` after the entire bootstrap procedure has finished, after the "damage" was already done.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+                session.execute("DROP KEYSPACE k")
+
+    async def stop(self) -> None:
+        """Stop a running server. No-op if not running. Uses SIGKILL to
```
Why don't we want graceful shutdown? Even if we want "hard kill" for some tests, we still want to test graceful shutdowns as well, no? (so `stop(self)` would be graceful, and we would have a `kill(self)` for hard kill). For graceful shutdown we could use SIGTERM.

nyh replied:
In 99.9% of the cases where I run a single test function, a graceful shutdown is a waste of time.
If shutdown is very fast (sub-second) and you usually run many tests and not just one (I guess this is typical of test.py) then I guess it won't hurt to have a graceful shutdown.

kostja replied:
We want both because there will be restarts. I have explained this already in a previous review.

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+        logging.info("Uninstalling server at %s", self.cfg["workdir"])
+
+        shutil.rmtree(self.tmpdir)
+        pathlib.Path(self.cfg["workdir"]).unlink(missing_ok=True)
```
The wrapping in `Path` is not needed (these are already Paths)

kbr- commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+            "cluster_name": cluster_name,
+            "host": None,
+            "seeds": seed,
+            "workdir": None,
```
It seems like you often use these keys inside the class below, so they should be member variables (to preserve the types for instance), and you could construct the dictionary when you actually need it from these member variables.

nyh commented on test/pylib/host_registry.py:
```diff
@@ -0,0 +1,82 @@
+            self.subnet = "127.{}.{}".format(random.randrange(1, 254),
+                                             random.randrange(0, 255))
+            self.lock_filename = os.path.join(os.getenv('TMPDIR', '/tmp'), "scylla-" + self.subnet)
+            self.lock_file = open(self.lock_filename, "w")
```
A reminder: this is Python, not C++. You don't have to agonize over this strong typing thing unless you force yourself to do it :-)

kostja commented on test/pylib/host_registry.py:
```diff
@@ -0,0 +1,82 @@
+                os.unlink(self.lock_filename)
+                self.lock_filename = ""
+
+        self.cleanup = cleanup
```
self is captured by the lambda, I return the lambda

kostja commented:
@nyh @kbr- Could you please send your comments over email? It's hard to keep track of what's going on here. Besides, I can see there is a lot of the new comments about places that didn't change since v3 or v2. Haven't you reviewed this patch already? Can I be certain that the code that doesn't hasn't received comments is not going to receive them going forward?

nyh commented:
> @nyh @kbr- Could you please send your comments over email?

If you want me to send comments over email only, not in pull requests, please do not open any pull requests. It's as simple as that. Moreover, if you look above, my only comment via this PR was about the last patch you added (with the tty thing) - and you forced-pushed this causing me to come here and comment here, not by email :-( Had you sent a mail instead, I could have replied there.

Finally, please adhere to Postel's law -  "be conservative in what you do, be liberal in what you accept from others". You should accept other people's review in the form that they sent it. @kbr- obviously spent a lot of time making a lot of very specific comments on your code - so now you can make the effort to read it where he sent it, and not ask him to re-send the same comments in a different form.

> It's hard to keep track of what's going on here.

As I told you and others before, this is **exactly** the problem with reviews over email and github pull requests solved (in fact, this is the only thing I like better in PRs than mails). The reviews of v1, ... v4 happened over mail, so now that you sent v5 (I didn't even see it by mail), nobody remembers what was reviewed in those versions, so any reviewer needs to start from scratch if he doesn't remember exactly what he - or others - said in the previous iterations.

Even if we don't **have** to start a review from scratch, e.g., because you claimed that you only changed on line of code from v4 to v5, you can't ask a reviewer to only review one this one line.... A reviewer puts a "stamp of approval" on your entire patch, not just the line that changed from v4 to v5. If @kbr- felt that he doesn't remember v4, or never reviewed v4, or suddenly remembers there was something in v4 he forgot to say - it's his right (and duty as a reviewer) to speak out.

One of the problems in your patch series is that it does a lot of different things, and any objection to any one of them delays it and caused you to send v2, v3, ... v5, v6, etc. One way you can avoid this in the future is to start sending smaller patch series, that do less. I recommended to you from the start (and you partially took this into account) that you don't need to overhaul all the test types, etc., in one huge patch series. You could have started just with the new Raft suite that you wanted to add, and later try to get the other suites in. For example, my long discussion with you about what the output of the tests look like would have been avoided if your initial patch series did not touch the cql-pytest suite (to be honest, I care less on what the debugging logs of the new Raft tests will look like - I will not be the one debugging them). Yes, this would have just delayed this inevitable discussion - we'll have it when you finally want to change the cql-pytest suite - but by then you will have at least got your first patch series in.

By the way, I don't think Kamil's comments about the types are new. I seem to remember that Kamil and I held opposite approaches: I was against any use of types in Python **testing** code, claiming it is a complete waste of time and makes a beautiful language C++-ugly (C++-ugly is like butt-ugly, just uglier ;-)). while Kamil was in favor of going all-in with using types. You can choose which of these two opinions you prefer - but by choosing to use types, you basically decided to accept comments about the way you used these types (and you'll only get these comments from Kamil, not me). When I look at Kamil's comments about the types, I feel happy that I rejected these types in my own test code and don't need to deal with those silly (in my opinion) issues - but you do use types, so you need to deal with comments about them.

> Besides, I can see there is a lot of the new comments about places that didn't change since v3 or v2. Haven't you reviewed this patch already? Can I be certain that the code that doesn't hasn't received comments is not going to receive them going forward?

You can't.

For the future, if you don't want many iterations, the best approach I can recommend is not to send very large patch series which will need more and more iterations of review. Another recommendation I can give is try to avoid mixing completely new features (which will see few objections) with changes to existing features (which will be heavily scrutinized by people who care about the existing features).

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+# it easier to restart. Sic: if you make a typo on the command line,
+# Scylla refuses to boot.
+SCYLLA_CMDLINE_OPTIONS = [
+    '--smp', '2',
```
One will need to send a patch to test.py to get it.

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+            await self.uninstall()
+
+        self.stop_artifact = stop_server
+        self.uninstall_artifact = uninstall_server
```
They lambdas capture self, class methods don't.

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -0,0 +1,307 @@
+import re
+import shutil
+import time
+import aiohttp
```
Yes it is

kostja commented on test.py:
```diff
@@ -293,6 +298,71 @@ def pattern(self):
         return "*_test.cql"
 
 
+class PythonTestSuite(TestSuite):
```
Adding type hints to test.py touches unrelated parts of the code, so I dropped it.

kostja commented on test.py:
```diff
@@ -293,6 +298,71 @@ def pattern(self):
+        # Suite artifacts are removed when
+        # the entire suite ends successfully. If it
+        # fails, we'd like to keep the data dir. This
+        # is why uninstall_server is not an exit artifact.
```
I rephrased the comment.

kbr- commented:
> Could you please send your comments over email? It's hard to keep track of what's going on here

It would be impossible to track anything if I commented a github PR over email. You could've sent the patchset over email if you wanted review over email. I'm not gonna write the comments a second time now

> Besides, I can see there is a lot of the new comments about places that didn't change since v3 or v2. Haven't you reviewed this patch already? 

No, I only asked you to add static types before so I could understand and properly review the code. Now I'm properly reviewing it

> Can I be certain that the code that doesn't hasn't received comments is not going to receive them going forward?

You probably never can

nyh commented:
@kostja can you please summarize what was new in the force-pushed test.py.v5 ?

kbr- commented:
@nyh 
https://github.com/scylladb/scylla/compare/6a0385a30cc1bbe0fae7c3054863ea8c2054bbcf..c9d843ecccde4bd9e739fef28b7b03e7a18a0d04

nyh commented on test/topology/conftest.py:
```diff
@@ -0,0 +1,25 @@
+    parser.addoption('--port', action='store', default='9042',
+                     help='CQL server port to connect to')
+    parser.addoption('--ssl', action='store_true',
+                     help='Connect to CQL via an encrypted TLSv1.2 connection')
```
A previous review asked why does this parser.option() is in test/topology/conftest.py, but the use of it is in test/conftest.py. What was the conclusion?

kostja replied:
I moved it. It's a weird thing of github pull requests is that  you begin reviewing before I finish, but as soon as I force-push. And I need to be force-pushing to save work.

kbr- replied:
You can force-push to another branch to save work, and only force-push to the PR-d branch after you finished with everything.

nyh commented on test/topology/test_null.py:
```diff
@@ -0,0 +1,26 @@
+    cql.execute("DROP TABLE " + table)
+
+
+def test_delete_empty_string_key(cql, table1):
```
I'm curious what this test is about... How is empty string test (which it seems you copied from cql-pytest?) related to the topology?
If is just a placeholder (?) and will be relaced later, or something, please explain in the commit message.

kostja replied:
This test is about nothing. I wrote about it before. It tests that clusters start and stop correctly.

kostja replied:
In other words, it tests suite.yaml

nyh commented on test/conftest.py:
```diff
@@ -1,9 +1,14 @@
+# fixtures directly from test/pylib).
+#
+# To enable, add addopts = --confcutdir .. to your pytest.ini
+#
```
Oh, here is the comment I asked for...
The first and last lines ("*all* tests) and "to enable, ...", are contradictory. It's not all tests if it's only those which enable it. Maybe better say something like:

Common pytest fixtures for all subdirectories which specifically request these
common pytest fixtures by adding "addopts = --confcutdir .." in their pytest.ini.

nyh commented on test/conftest.py:
```diff
@@ -12,6 +17,18 @@
+                     help='CQL server port to connect to')
+    parser.addoption('--ssl', action='store_true',
+                     help='Connect to CQL via an encrypted TLSv1.2 connection')
+
```
Now I see you moved this code in this patch, which is confusing to review.
Anyway, my comment in the original place is still relevant - the name "--port" and "--ssl" make sense for CQL tests. Other types of tests will probably need different parameters, and the very general-sounding name "--port" can get confusing.

nyh commented on test/pylib/cql_repl/cql_repl.py:
```diff
@@ -0,0 +1,64 @@
+                    ofile.write("\n")
+            except Exception as e:
+                # Avoid host name in the message - make output stable
+                msg = re.sub(r"127\.\d+\.\d+\.\d+", "127.0.0.1", str(e))
```
Why assume the host name is always 127.*.*.*. Can't it be set with the "--host" option to something else, e.g., when you're testing a remote or manually-run Scylla?

kostja replied:
ok

nyh reviewed (changes requested):
Looks mostly good to me, with some minor comments.

I still have no idea how Scylla logs will be visible when looking at cql-pytest failures  . The test output no longer contains the Scylla log (and you did it deliberately) - instead the Scylla logs is a vailable in files. The question is how the user will find the file, and how easy it is to do on Jenkins. I'm still waiting for you to say "yes, I tried this and it's easy, I just did ... to find the log file". Since I myself didn't try it, I don't know.
You change to "--verbose" in one of the patches is not relevant to this question.

kostja commented on test/conftest.py:
```diff
@@ -0,0 +1,86 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+# This file configures pytest for all tests in this directory, and also
+# defines common test fixtures for all of them to use.
```
If pytest.ini is missing, both get used. If pytest.ini is present, it has to specify which ones get used. The order is oustide-in.

kostja replied:
I added the logging.

kostja commented on test/conftest.py:
```diff
@@ -0,0 +1,86 @@
+
+# "cql" fixture: set up client object for communicating with the CQL API.
+# The host/port combination of the server are determined by the --host and
+# --port options, and defaults to localhost and 9042, respectively.
```
Sure, right now the conftest.py  in test/ is shared among topoloy, cql-pytest and cql/ suites. It is a trivial patch to "reconsider" it when more ports are needed (if this happens).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/198/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/199/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/212/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/219/

benipeled commented:
CI failure:
```
21:31:36  Error from server: code=2200 [Invalid query] message="Role cassandra doesn't exist."
21:31:36  Traceback (most recent call last):
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 1085, in <module>
21:31:36      sys.exit(asyncio.run(main()))
21:31:36    File "/usr/lib64/python3.9/asyncio/runners.py", line 44, in run
21:31:36      return loop.run_until_complete(main)
21:31:36    File "/usr/lib64/python3.9/asyncio/base_events.py", line 647, in run_until_complete
21:31:36      return future.result()
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 1056, in main
21:31:36      await run_all_tests(signaled, options)
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 930, in run_all_tests
21:31:36      await reap(done, pending, signaled)
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 912, in reap
21:31:36      result = coro.result()
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 170, in run
21:31:36      await test.run(options)
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 617, in run
21:31:36      async with self.suite.clusters.instance() as cluster:
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/pool.py", line 52, in __aenter__
21:31:36      self.obj = await self.pool.get()
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/pool.py", line 36, in get
21:31:36      await self.pool.put(await self.build())
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 344, in start_simple
21:31:36      await server.install_and_start()
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_server.py", line 124, in install_and_start
21:31:36      await self.start()
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_server.py", line 279, in start
21:31:36      await self.force_schema_migration()
21:31:36    File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_server.py", line 292, in force_schema_migration
21:31:36      session.execute("CREATE KEYSPACE k WITH REPLICATION = {" +
21:31:36    File "/usr/local/lib64/python3.9/site-packages/cassandra/cluster.py", line 2668, in execute
21:31:36      return self.execute_async(query, parameters, trace, custom_payload, timeout, execution_profile, paging_state, host, execute_as).result()
21:31:36    File "/usr/local/lib64/python3.9/site-packages/cassandra/cluster.py", line 4953, in result
21:31:36      raise self._final_exception
21:31:36  cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Role cassandra doesn't exist."
```

kostja commented:
Pushed a fix.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/227/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/224/

nyh commented on test/pylib/scylla_server.py:
```diff
@@ -215,9 +216,18 @@ async def cql_is_up(self) -> bool:
+                    # work, so rely on this "side effect".
+                    session.execute("CREATE KEYSPACE k WITH REPLICATION = {" +
+                                    "'class' : 'SimpleStrategy', 'replication_factor' : 1 }")
+                    session.execute("DROP KEYSPACE k")
```
Is the keyspace creation/deletion somehow needed for testing the "cassandra" role?
In my code, I used the silly request "use system" as a do-nothing request. But maybe that doesn't fail without the role?

Anyway, it seems to me that you actually discovered a Scylla bug, and we shouldn't paper over it with a test fix (although we can - temporarily).

kostja replied:
If you look at the test error log, force_schema_migration() fails on partially initialized instances. force_schema_migration() uses create/drop keyspace. This is why I use the same statements. I haven't checked for a minimal statement that would trigger the problem. use is only calling validate_login() so not necessarily checking the role.

kostja commented:
https://jenkins.scylladb.com/job/releng/job/Scylla-CI/243/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/243/

kbr- commented:
@nyh are you satisfied? Let's queue this thing!

kostja commented:
> @kostja can you please summarize what was new in the force-pushed test.py.v5 ?

What is this pull request waiting for?

kbr- commented:
PING
@nyh please give your blessing or post a review

nyh commented:
@kbr-  my review from 11 days ago was:

>Looks mostly good to me, with some minor comments.
>I still have no idea how Scylla logs will be visible when looking at cql-pytest failures . The test output no longer contains the Scylla log (and you did it deliberately) - instead the Scylla logs is a vailable in files. The question is how the user will find the file, and how easy it is to do on Jenkins. I'm still waiting for you to say "yes, I tried this and it's easy, I just did ... to find the log file". Since I myself didn't try it, I don't know. Your change to "--verbose" in one of the patches is not relevant to this question.

Kostja still hasn't answered this question about the log files, and confirming that he tried to run a failing test on Jenkins and see what it looks like.

But if you want to commit this series, please go ahead - I'm not against. If we'll see problems with this going forward, we can always fix those problems then. If we have cql-pytest problems and I can't find the log files, I'll ask Kostja for help finding them :-)

kbr- commented:
@kostja please do the following so we can proceed:
1. setup one of the tests to fail artificially (preferably two tests: a single-node and a cluster test).
2. Run the Jenkins CI. Wait for it to fail.
3. Find the Scylla logs in the Jenkins artifacts. Post the URL.
4. Give an instruction how to find, in general, the logs starting from a CI link given by the CI bot.
5. (Preferably, but optionally) attach the instruction in some documentation file somewhere in the test directory.

kbr- commented:
Queued

nyh commented:
> Dequeued, there was still some problem with collecting the logs (apparently, we only collect the logs when the `-s`/`--save-log-on-success` flag is provided, even for failed tests)

This looks like an outright bug, but beyond that I think it reiterates my proposal (which, like Cato the Elder, I repeated several times ;-)),  that test output should NOT look like this: https://jenkins.scylladb.com/job/scylla-master/job/next/4859/testReport/junit/(root)/non-boost%20tests/Test___cassandra_cql_test_dev_6102/

The full content of that link is:
> Test pytest test/pylib/cql_repl/cql_repl.py --host=127.58.179.6 --input=test/cql/cassandra_cql_test.cql --output=/jenkins/workspace/scylla-master/next/scylla/testlog/dev/cassandra_cql_test.6102.reject failed, check the log at /jenkins/workspace/scylla-master/next/scylla/testlog/dev/cassandra_cql_test.6102.log

wouldn't it be more useful to insert the full content of those "log" and "reject" files instead of just their names - which then require an inconvenient lookup on Jenkins?

kbr- commented:
@nyh do you know how Jenkins constructs the content of this site (what files it uses etc.)?

nyh commented:
> You can't provision for inclusion of everything you might need - this is equivalent to concatenating entire testlog dir into the log output.

I didn't say "everything". You yourself thought that specific log files will be interesting so you added pointers to them. You could have, instead, put the interesting information in the test output (the xml) and not in a separate file.


> Failing tests should leave artifacts which should be available for inspection: https://jenkins.scylladb.com/job/scylla-master/job/next/4859/artifact/testlog/x86_64_debug/lwt_batch_test.6141.log I will take a look why server logs are missing there. I don't expect it to help though. The branch went through numerous CI rebuilds before it was queued - it's not likely that the next rebuild will fail. But we'll see.

There have already been 4 runs that failed with the same two tests. It's not a fluke. But I have no idea what it is because I can't see the logs.

Please try checking out "next" and running it yourself with this PR on top of it. It's quite possible that some other patch changed something in the CQL output. Remember that your cql-repl tests are sensitive to any changes in the output, and your patch replaced all of the output with output that you saw - so maybe one of these outputs are no longer correct.

kostja commented:
* nyh ***@***.***> [22/05/11 17:30]:
> > You can't provision for inclusion of everything you might need - this is equivalent to concatenating entire testlog dir into the log output.
> 
> I didn't say "everything". You yourself thought that specific log files will be interesting so you added pointers to them. You could have, instead, put the interesting information in the test output (the xml) and not in a separate file.

Putting them to xml is not the same as putting them to stdout. so
you'd like to put these logs into xml?

> > Failing tests should leave artifacts which should be available for inspection: https://jenkins.scylladb.com/job/scylla-master/job/next/4859/artifact/testlog/x86_64_debug/lwt_batch_test.6141.log I will take a look why server logs are missing there. I don't expect it to help though. The branch went through numerous CI rebuilds before it was queued - it's not likely that the next rebuild will fail. But we'll see.
> 
> There have already been 4 runs that failed with the same two tests. It's not a fluke. But I have no idea what it is because I can't see the logs.
> 
> Please try checking out "next" and running it yourself with this PR on top of it. It's quite possible that some other patch changed something in the CQL output. Remember that your cql-repl tests are sensitive to any changes in the output, and your patch replaced all of the output with output that you saw - so maybe one of these outputs are no longer correct.
> 

I have run it locally, it runs fine, so it's not the output.
Please also follow the link I sent in the previous reply, it tells
what it is - scylla server disappears in the middle of the test or
fails to start.

-- 
Konstantin Osipov, Moscow, Russia

nyh commented:
> I have run it locally, it runs fine, so it's not the output. Please also follow the link I sent in the previous reply, it tells what it is - scylla server disappears in the middle of the test or fails to start.

I saw this, I don't know if this is the only problem (you now have multiple runs that failed on these two tests, you can check them all).
The Scylla server failing to start can't explains the consistent failure of the same two tests!
What can explain this is if Scylla crashed in the middle of a specific test (but I don't know if this is consistent with the error message you saw). test/cql-pytest/run had (and still has) special handling for this case - first we check _after_ each test if Scylla is still alive (and if it isn't, we print a clear error), and second, on failure I dumped the full Scylla log to the stdout (i.e., consoleText), so we'll know exactly why it crashed.

Before I added those things to test/cql-pytest/run, it was hard to understand what happened when a test crashed Scylla - instead of a report that test X failed, we got a report that the next test Y failed (because it couldn't start), and we didn't know why we cause we didn't have the Scylla log.

nyh commented:
> Putting them to xml is not the same as putting them to stdout. so you'd like to put these logs into xml?

As an example, take a look at what happens for failed Boost tests. For example, https://jenkins.scylladb.com/job/scylla-master/job/next/4863 the test boost.mutation_test.debug.test/boost/mutation_test.test_query_digest failed. The entire test output appears in the XML file which is conveniently formatted and linked directly on the test page:

https://jenkins.scylladb.com/job/scylla-master/job/next/4863/testReport/junit/boost.mutation_test.debug/test_boost_mutation_test/test_query_digest/

The same full test output is **also** on the standard output at https://jenkins.scylladb.com/job/scylla-master/job/next/4863/console

Why is it so unthinkable that a failed cql-repl tests have the same level of information in those two convenient locations instead of being stored in a separate file that is inconvenient to find (and currently, is worse - it isn't even available!)

I'm not saying that we need the same information in *both* locations, but I think even that doesn't hurt. In test/cql-pytest/run I had different information in the two locations - in the xml file there was the pytest output, and in the stdout there was the full Scylla log (which can be useful if, for example, Scylla crashed). The specific split isn't very important. But the important point is that both the (formatted version of the) xml file and the stdout are very convenient to access in Jenkins. Accessing random files is simply much less convenient - if at all possible (as you saw recently with the cql-repl failures, it simply isn't possible currently, because those files weren't saved because of some other bug).

nyh commented:
> > Putting them to xml is not the same as putting them to stdout. so you'd like to put these logs into xml?
> 
> As an example, take a look at what happens for failed Boost tests. For example, https://jenkins.scylladb.com/job/scylla-master/job/next/4863 the test boost.mutation_test.debug.test/boost/mutation_test.test_query_digest failed. The entire test output appears in the XML file which is conveniently formatted and linked directly on the test page:
> 
> https://jenkins.scylladb.com/job/scylla-master/job/next/4863/testReport/junit/boost.mutation_test.debug/test_boost_mutation_test/test_query_digest/
> 
> The same full test output is **also** on the standard output at https://jenkins.scylladb.com/job/scylla-master/job/next/4863/consoleText

A small correction - it appears this isn't **exactly** the same output. The individual test page ("xml") contains the result of this specific test only. The consoleText ("stdout") actually contains a bit more - the entire run of the test program. This includes various messages - such as a printout of the random key - which aren't in the xml file. But my main idea remains the same - the test output is in those two files, not in some separate file whose filename is printed in those files.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/328/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/330/

kostja commented:
kostja@hulk:~/work/scylla/scylla/testlog$ diff -u  ../test/cql/cassandra_cql_test.result ./release/cassandra_cql_test.583.reject                  
--- ../test/cql/cassandra_cql_test.result       2022-05-11 17:21:10.244679691 +0300                                                               
+++ ./release/cassandra_cql_test.583.reject     2022-05-12 23:10:31.743293910 +0300                                                               
@@ -14,7 +14,7 @@                                                                                                                                 
 > -- setup                                                                                                                                       
 > CREATE KEYSPACE ks1 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}                                                    
 >         AND DURABLE_WRITES = true;                                                                                                             
-OK                                                                                                                                               
+Keyspace 'ks1' already exists                                                                                                                    
 > ALTER KEYSPACE ks1 WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1' : 1 };

nyh commented:
Good, this explanation makes sense (but I don't understand how it fits with your earlier explanation of "server didn't start" - did we have two separate problems?).

All test/cql-pytest tests are designed to run on a shared Scylla, so they always use random keyspace and table names, not names like "ks1", and orthogonally, also delete the keyspaces/tables they create at the end of the test. 
So it looks like if cql-repl tests use non-random keyspace/table names, in tests that can't use IF NOT EXIST to create the keyspace, you'll just need to use a unique name for it?

kostja commented:
We observe "the server failed to start" for the very same test when run in --debug mode. So it is at least somewhat different problem -we get different diagnostics in debug mode, whereas we should get the same. I am investigating the debug case separately.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/346/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/353/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/354/

nyh commented:
> OK, so I tried 1) capturing the start exception b) capturing the logs. This leads to a massive bloat of the console ouptut. Imagine you fail to start the server. It's going to lead to one test failing after another, all with the same issue.

Remember when you started this pytest work I told you that cql-pytest/run already has two years of experience, isn't as bad as you thought and you're bound to rediscover many of the problems it solved? This is an example of what I meant :-)

When test/cql-pytest/run failed to run Scylla, the failed boot log appears on the stdout (consoleText in Jenkins) - just once. None of the individual tests got run, or had any results in their xmlfiles - neither failed nor successful. You didn't get the same log copied a hundred times, just once. Jenkins would fail because the whole test/cql-pytest/run failed (this outside run has a test name of its own, by the way, separate from the individual tests). The Scylla boot failure appeared in stdout and was helpful in diagnosing what happened.

Moreover, as I noted before, in other cases Scylla did boot, but then crashed in the middle of some test (this is one of the downsides of running many tests on the same server - as you rightly pointed out in the past). In the early days of cql-pytest, we would then get a hundred other tests claiming to "fail" because they failed to send any request to Scylla. I solved this by adding a verification step after each test (done via an autouse fixture) that Scylla is still working. If it doesn't, the entire pytest is aborted - we don't proceed to trying and obviously-failing any more additional tests. We then report exactly which test was running when Scylla crashed. In your implementation, you can do something else - perhaps send the following tests to another Scylla server (or start another one). But you'll need to do something.

None of this is critical for the first release, however. test/cql-pytest/run also didn't have all these features when it's just started. As long as I can still run test/cql-pytest/run myself (this is why I insisted on it), I don't mind a temporary regression in the usefulness of the test framework - at least if you're willing to incrementally improve it later.

nyh commented:
> Wouldn't it be enough to just
> 
>     1. save the server's log in its own file,
>     2. have the test.py console log _point_ to the server log, instead of writing it inline?

I still think that all these separate files are a mistake. I believe that the only set of files (besides the stdout) that we should have are the XML files which Jenkins makes conveniently visible in its "test results" output. I think that making the test results a single line saying "there's nothing to see here, please go to /some/where/another/file.txt" is a user-experience mistake.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/381/

kostja commented:
I updated the branch, please review the changes before I squash them.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/391/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/394/

kbr- commented:
Looks like an assertion failed in scylla_server.py:
```
[2022-05-18T01:12:50.922Z] Traceback (most recent call last):
[2022-05-18T01:12:50.922Z]   File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 1122, in workaround_python26789
[2022-05-18T01:12:50.923Z]     code = await main()
[2022-05-18T01:12:50.923Z]   File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 1089, in main
[2022-05-18T01:12:50.923Z]     await run_all_tests(signaled, options)
[2022-05-18T01:12:50.923Z]   File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 961, in run_all_tests
[2022-05-18T01:12:50.923Z]     await reap(done, pending, signaled)
[2022-05-18T01:12:50.923Z]   File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 943, in reap
[2022-05-18T01:12:50.923Z]     result = coro.result()
[2022-05-18T01:12:50.923Z]   File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 177, in run
[2022-05-18T01:12:50.923Z]     await test.run(options)
[2022-05-18T01:12:50.923Z]   File "/jenkins/workspace/releng/Scylla-CI/scylla/./test.py", line 649, in run
[2022-05-18T01:12:50.923Z]     self.args.insert(0, "--host={}".format(cluster[0].host))
[2022-05-18T01:12:50.923Z]   File "/jenkins/workspace/releng/Scylla-CI/scylla/test/pylib/scylla_server.py", line 374, in __getitem__
[2022-05-18T01:12:50.923Z]     assert(self.start_exception is None)
[2022-05-18T01:12:50.923Z] AssertionError
[2022-05-18T01:12:51.736Z] Exception in thread Task Scheduler:
[2022-05-18T01:12:51.736Z] Traceback (most recent call last):
[2022-05-18T01:12:51.736Z]   File "/usr/lib64/python3.9/threading.py", line 973, in _bootstrap_inner
[2022-05-18T01:12:51.736Z]     self.run()
[2022-05-18T01:12:51.736Z]   File "/usr/local/lib64/python3.9/site-packages/cassandra/cluster.py", line 4236, in run
[2022-05-18T01:12:51.736Z]     future = self._executor.submit(fn, *args, **kwargs)
[2022-05-18T01:12:51.736Z]   File "/usr/lib64/python3.9/concurrent/futures/thread.py", line 167, in submit
[2022-05-18T01:12:51.736Z]     raise RuntimeError('cannot schedule new futures after shutdown')
```

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/398/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/403/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/409/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/413/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/449/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/448/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/454/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/461/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/465/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/468/

nyh commented on test.py:
```diff
@@ -63,6 +64,10 @@ class palette:
     diff_mark = create_formatter(colorama.Fore.MAGENTA)
     warn = create_formatter(colorama.Fore.YELLOW)
     crit = create_formatter(colorama.Fore.RED, colorama.Style.BRIGHT)
+    ansi_escape = re.compile(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])')
```
I guess you tested this function, but I can't readit and wonder how it matches an ANSI color change sequence like `ESC[01;34m`.

kostja replied:
Yes, I did test it, it removes the escape sequences added by colorama. I think you are mixing up ESC as three letters - E - S - C and ESC as a key on the terminal keyboard with an assigned ANSI code which is 1B, that is 27 in decimal. The escape sequence starts with ESC control code, not ESC letters.

nyh commented on test.py:
```diff
@@ -517,53 +517,78 @@ def __init__(self, test_no, shortname, suite):
+        async with self.suite.clusters.instance() as cluster:
+            self.args.insert(1, "--host={}".format(cluster[0].host))
+            # If pre-check fails, e.g. because Scylla failed to start
+            # or crashed between two tests, fail entire test.py
```
If Scylla fails between two tests, it's enough to do a "post check" (right after a test finished, were you can also report which test is to blame). You don't really need to do a "pre-check" yet again a millisecond later when starting the next test.

You only need to check that Scylla booted correctly when starting Scylla, not before each test.

Unless I'm misunderstanding what you're trying to do here.

kostja replied:
I agree we could optimize out getting the number of keyspaces - we don't need to get it each time in the beginning of the test. pre_check(), however, also prints a line into the log file with the test name. I think I will rename pre_check to before_test and post_check to after_test(), these names would better reflect the responsibility of these two calls.

Finally, you maybe are raising in this comment the issue of when to throw if the server fails to start. I thought that it is perhaps better to throw from pre_check (before_test()) since then the test can consider what to do with it. If building a new server throws, i.e.. when the server is being added to the pool, I'll have to handle this twice - first when the server fails to start, and then when it crashes after the test.

To sum up:
- I will rename the two functions
- I will optimize querying the canonical keyspace count so that it's done once per server, not once for each test.
- I will keep the before- function 
- I will keep suppressing the start exception and will raise it in the before- function.

Please let me know what you think

nyh commented on test.py:
```diff
@@ -517,53 +517,78 @@ def __init__(self, test_no, shortname, suite):
+                    set_summary("failed: no output file")
+                elif not os.path.isfile(self.result):
+                    set_summary("failed: no result file")
+                    self.is_new = True
```
I see this is pre-existing code, but I still wonder: why set is_new=True only in this specific "no result file" case?

kostja replied:
"is_new" is there to support the workflow of adding new tests. You write test CQL, run the test, and commit the produced result file. If there is a result file, then there is a canonical result, so the test is not new - in other words, it's not the first time we run this test.

nyh commented on test.py:
```diff
@@ -517,53 +517,78 @@ def __init__(self, test_no, shortname, suite):
+                # Server log bloats the output if we produce it in all
+                # cases. So only grab it when it's relevant:
+                # 1) failed pre-check, e.g. start failure
+                # 2) failed test execution.
```
I think you should check **post**-check, not **pre**-check.
In other words, if Scylla crashed during a specific test - even if the test itself passed miraculously (?) - we should report this failure for **that** test. I don't want to see a failure report for the **next** test just because when it started, its pre-check failed. That is very confusing - you see a failure in test "xyz" but after a lot of thinking you realize there is nothing wrong with test xyz, it was actually the test abc that ran just before it crashed Scylla, and xyz couldn't even start.

kostja replied:
If scylla crashed during the test, post_check() will detect it because it runs a query against scylla - getting the keyspace count. This will be correctly attributed to the test which caused the crash.

kbr- commented on test/pylib/artifact_registry.py:
```diff
@@ -28,22 +28,22 @@ def __init__(self) -> None:
         logging.info("Cleaning up before exit...")
-        for name in self.suite_artifacts:
-            for artifact in self.suite_artifacts[name]:
+        for suite_key in self.suite_artifacts:
```
Better yet, since you're interested in the values and not the keys:
```python
for artifacts in self.suite_artifacts.values():
```
then use `artifacts` instead of `self.suite_artifacts[suite_key]`.

kostja replied:
done

kbr- commented on test/pylib/artifact_registry.py:
```diff
@@ -23,30 +23,42 @@ class ArtifactRegistry:
     def __init__(self) -> None:
-        self.suite_artifacts: Dict[Suite, List[Artifact]] = {}
-        self.exit_artifacts: List[Artifact] = []
+        self.suite_artifacts: Dict[str, List[Artifact]] = {}
```
Why change the key to `str`? It was IMO much more readable/understandable when you used `Suite`.

kostja replied:
because I needed __exit__ key for exit artifacts that do not belong to any suite. In Lua None (nil) can not be a dict key type, so I carried over that habit. I don't have a strong opinion about what is better in Python, using strings or None  + objects, so I'll just do it the way you suggest. I do consider, however, comments which point out bugs in my code more valuable than style comments on bugfixes I made without a reviewer request...

kostja replied:
done

kbr- commented on test/pylib/artifact_registry.py:
```diff
@@ -23,30 +23,42 @@ class ArtifactRegistry:
-        self.suite_artifacts: Dict[Suite, List[Artifact]] = {}
-        self.exit_artifacts: List[Artifact] = []
+        self.suite_artifacts: Dict[str, List[Artifact]] = {}
+        self.exit_artifacts: Dict[str, List[Artifact]] = {}
```
Here you can use `Optional[Suite]` as the key type. In `add_exit_artifact` you can then use `suite` directly as the key. No need to use a special `__exit__` string.

kostja replied:
ok

kbr- commented on test.py:
```diff
@@ -648,11 +650,22 @@ def print_summary(self):
+                self.is_post_check_ok = True
+                self.success = code
+            except Exception as e:
                 self.server_log = cluster[0].read_log()
```
The previous version saved `self.server_log` in case `not self.success` was true. Right now it only saves it in case of Exception. It may happen that no exception is thrown but `self.success == False` for example (unless I'm mistaken?). Don't we want to save the log in that case too?

kostja replied:
We always save the server log in testlog/. Here we save the relevant lines of the log to print it to stdout. Once again, the matter of having things printed to stdout is highly subjective. I spent quite a bit of time debugging things with CI and locally, and think that what I did is best. When the test itself fails, (return code is not 0) there is most likely noting wrong with the server, so we 're more interested in the test log, which we print, not server log, which will clutter the ouptut.

kostja replied:
(resolving without further action)

kostja commented:
The current branch, before addressing review comments and squashing, is saved as test.py.v6

kostja commented on test/pylib/scylla_server.py:
```diff
@@ -385,7 +390,7 @@ def _get_keyspace_count(self) -> int:
         return keyspace_count
 
-    def pre_check(self) -> None:
+    def pre_check(self, name) -> None:
```
I will rename them to avoid making an impression that they don't have side effects.

kostja replied:
done

kostja commented:
@kbr- @nyh we have another problem with this pull request, pointed out by @benipeled . right now I create server tmpdirs in TMPDIR, and make a symlink to testlog/. The issue is, s3 doesn't support symlinks and hardlinks don't work across filesystems. So how about I drop this entire optimization of using tmpdir and create server subdirs in testlog/ so that @benipeled can easily capture this stuff in CI?

kbr- commented:
> @kbr- @nyh we have another problem with this pull request, pointed out by @benipeled . right now I create server tmpdirs in TMPDIR, and make a symlink to them from testlog/. The issue is, s3 doesn't support symlinks and hardlinks don't work across filesystems (the whole point of using TMPDIR is that it is tad faster because it's tmpfs). So how about I drop this entire optimization of using tmpdir and create server subdirs in testlog/ so that @benipeled can easily capture this stuff in CI?

Wouldn't it be possible to simply move everything from tmpdir into `testlog/` after we're done with the testing?
Another idea would have the CI collect stuff directly from tmpdir, if that stuff is there. Maybe CI shouldn't have any directory hardcoded, but obtain it from the test harness. The harness could use tmpdir, `testlog/` or whatever.

kbr- commented:
But the tmpdir optimization did not exist before this PR, correct? (if it did, you the problem you describe must have been solved already) Then indeed the best path right now would be to not introduce it in this patchset.

kbr- commented:
According to the output you posted, `test.py` saves in `./testlog/` by default, it doesn't use the actual (in-memory) temp dir.

benipeled commented:
> run.py uses tmpdir but doesn't have the goal to save scylla logs in CI. ./test.py already supports providing an alternative path:
> 
> ```
>   --tmpdir TMPDIR       Path to temporary test data and log files. The data
>                         is further segregated per build mode. Default:
>                         ./testlog.
> ```

@nyh seems like the option we discussed already exists,

@kostja the plan is to use `--tmpdir=<arch>_testlog` (or a similar pattern) on Jenkins and later - in case of failure - archive **all** content under the tmpdir. With such a change we'll not have to deal with specific-files refs.

kostja commented:
Fine, I will remove the symlinking from --tmpdir to getenv('tmpdir') then.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/540/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/572/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/581/
